api.go

```diff
@@ -0,0 +1,169 @@
+    in the Software without restriction, including without limitation the rights
+    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in all
+    copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+
+*/
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"net/http/httptrace"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// RunSimpleTest starts an HTTP/HTTPS API test of a site within crabby.  It uses Go's built-in net/http client.
+func RunApiTest(ctx context.Context, j Job, storage *Storage, client *http.Client, responses map[string]*JobResponse) {
+	for ; !prerequisitesMet(j, responses); {
+		log.Printf("Prerequisites not met for %s, going to sleep.\n", j.Name)
+		time.Sleep(1 * time.Second)
+	}
+	log.Printf("Prerequisites met. Running %s.\n", j.Name)
+
+	var method = strings.ToUpper(j.Method)
+	if method == "" {
+		method = http.MethodGet
+	}
+
+	req, err := http.NewRequest(method, j.URL, strings.NewReader(j.Body))
+	if err != nil {
+		log.Printf("unable to create request: %v", err)
+		return
+	}
+
+	addHeaders(req, j)
+
+	var t0, t1, t2, t3, t4 time.Time
+
+	trace := &httptrace.ClientTrace{
+		DNSStart: func(_ httptrace.DNSStartInfo) { t0 = time.Now() },
+		DNSDone:  func(_ httptrace.DNSDoneInfo) { t1 = time.Now() },
+		ConnectStart: func(_, _ string) {
+			if t1.IsZero() {
+				// connecting to IP
+				t1 = time.Now()
+			}
+		},
+		ConnectDone: func(net, addr string, err error) {
+			if err != nil {
+				log.Printf("unable to connect to host %v: %v", addr, err)
+			}
+			t2 = time.Now()
+		},
+		GotConn:              func(_ httptrace.GotConnInfo) { t3 = time.Now() },
+		GotFirstResponseByte: func() { t4 = time.Now() },
+	}
+
+	// We'll use our Context in this request in case we have to shut down midstream
+	req = req.WithContext(httptrace.WithClientTrace(ctx, trace))
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Failed to read response:", err)
+		return
+	}
+
+	// Send our server response code as an event
+	storage.EventDistributor <- j.makeEvent(resp.StatusCode)
+
+	// Save response for use by other jobs
+	if _, ok := responses[j.Name]; ok {
+		responses[j.Name].Completed = true
+		responses[j.Name].Response = resp
+		log.Println("Response saved job:", j.Name)
+		go waitAndClearResponse(responses, j.Name)
```

Comment:
That's not what I was going for here. Consider this: there are 2 jobs, A and B. B cannot run until A has run.

Since both of them live in go routines, they can occur in any order. This is intended to safeguard against the following situation: A executes, B executes, 10 seconds later B executes before A and succeeds (which shouldn't happen because B needs the response of A).

If I put this into a defer, this could happen: B tries to execute, A executes and finishes execution, B tries again to execute but fails because the data has already been cleared.

Reply:
As we discussed, this is definitely a good reason for running them synchronously within a single function.  Coordinating between routines is a nightmare and fraught with race conditions and leaks.